### PR TITLE
Add support for new endpoints.json format and make PartitionEndpointProvider the default

### PR DESCRIPTION
**Pull request by jeskew**
This PR adds support for the new endpoints file format.

/cc @mtdowling @cjyclaire @chrisradek @xibz

### CODE REVIEW

**Comment by mtdowling** [comment]
We talked about this PR in person. Let me know when this has changed and is ready for another review.

**Comment by jeremeamia** [comment]
This looked "fun" to do. ;-)

**Comment by jeskew** [comment]
@mtdowling I believe I covered what we discussed, and this PR is ready for another review.

File: src/ClientResolver.php [hunk 1]
```diff
@@ -412,6 +427,12 @@ public static function _apply_endpoint_provider(callable $value, array &$args)
             if (isset($result['signatureVersion'])) {
                 $args['config']['signature_version'] = $result['signatureVersion'];
             }
+            if (isset($result['signingRegion'])) {
+                $args['config']['signing_region'] = $result['signingRegion'];
```
Comment (mtdowling) [inline src/ClientResolver.php]: Is the user provides an explicit signing region in the client constructor, it should override what comes back from the endpoint provider. Is that case handled?

File: src/CommandPool.php [hunk 2]
```diff
@@ -58,7 +58,7 @@ public function __construct(
                 if ($before) {
                     $before($command, $key);
                 }
-                yield $client->executeAsync($command);
+                yield $key => $client->executeAsync($command);
```
Comment (mtdowling) [inline src/CommandPool.php]: Is this an unrelated change?

File: src/Endpoint/Partition.php [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    /**
+     * Determine if this partition contains the provided region.
+     *
+     * @param string $region
+     *
+     * @return bool
+     */
+    public function matchesRegion($region)
+    {
+        if (isset($this->definition['regionRegex'])) {
```
Comment (mtdowling) [inline src/Endpoint/Partition.php]: Couldn't you first check if the list of regions for the partition matches the given region, and only then fall back to regex based results?

File: src/Endpoint/Partition.php [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        $data += isset($this->definition['defaults'])
+            ? $this->definition['defaults']
+            : [];
+
+        return $data;
+    }
+
+    private function getSignatureVersion(array $data)
+    {
+        static $supportedBySdk = [
```
Comment (mtdowling) [inline src/Endpoint/Partition.php]: This is good, but on a side note: I wonder if we should add "sdkPreferredSignatureVersion" or something similar to the config file to remove this from each implementing SDK.

File: src/Endpoint/PartitionProviderInterface.php [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+interface PartitionProviderInterface
+{
+    /**
+     * Gets the partition containing a given region.
+     * 
+     * @param string $region
+     *
+     * @return Partition
+     */
+    public function getPartition($region);
```
Comment (mtdowling) [inline src/Endpoint/PartitionProviderInterface.php]: In order to allow this interface to be more extensible, perhaps this should be a more targeted method name... Like getPartitionFromRegion($region) or something.

File: tests/CommandPoolTest.php [hunk 6]
```diff
@@ -35,6 +35,19 @@ public function testEnsuresBeforeIsCallable()
         new CommandPool($client, [], ['before' => 'foo']);
     }
 
+    public function testYieldsKey()
```
Comment (mtdowling) [inline tests/CommandPoolTest.php]: Unrelated change?

File: src/CommandPool.php (hunk 2 above)
Comment (jeskew) [inline src/CommandPool.php]: Yes, but the `mapRegions` function doesn't make much sense without it. If the iterator provided to the command pool yields meaningful keys, they are only currently passed to the `before` callback, not `fulfilled` or `rejected`.

File: src/ClientResolver.php (hunk 1 above)
Comment (jeskew) [inline src/ClientResolver.php]: I believe that case would be handled, but I agree that there should be an explicit test for overriding signing_region and signing_name.

File: src/MultiRegionClient.php [hunk 7]
```diff
@@ -45,6 +51,25 @@ public static function getArguments()
...
+                'doc'     => 'AWS partition to connect to. Valid partitions include "aws," "aws-cn," and "aws-us-gov."',
+                'default' => function (array $args) {
+                    $region = isset($args['region']) ? $args['region'] : '';
+                    return PartitionEndpointProvider::defaultProvider()
+                        ->getPartition($region);
+                },
+                'fn'      => function ($value, array &$args) {
+                    if (is_string($value)) {
+                        $value = PartitionEndpointProvider::defaultProvider()
+                            ->getPartition($value);
```
Comment (mtdowling) [inline src/MultiRegionClient.php]: Isn't this trying to get a partition by partition name, but the method only works with region names?

File: src/MultiRegionClient.php [hunk 8]
```diff
@@ -113,6 +139,52 @@ public function getEndpoint()
     }
 
     /**
+     * @param $commandName
```
Comment (mtdowling) [inline src/MultiRegionClient.php]: Without a description, it's not clear what this method does.

File: src/MultiRegionClient.php [hunk 9]
```diff
@@ -113,6 +139,52 @@ public function getEndpoint()
...
+        $commandGen = function () use ($regions, $commandName, $args) {
+            foreach ($regions as $region) {
+                yield $region => $this->getClientFromPool($region)
+                    ->getCommand($commandName, $args);
+            }
+        };
+        $resultHandler = static function ($result, $region) use (&$results) {
+            $results[$region] = $result;
+        };
+        $pool = new CommandPool($this, $commandGen(), [
```
Comment (mtdowling) [inline src/MultiRegionClient.php]: I wonder if exposing any of the CommandPool options is something we should support? What if someone doesn't want to do these operations concurrently or wants to control the concurrency level?

File: src/MultiRegionClient.php [hunk 10]
```diff
@@ -45,6 +51,25 @@ public static function getArguments()
                     };
                 },
             ],
+            'partition' => [
```
Comment (mtdowling) [inline src/MultiRegionClient.php]: Can you add tests for configuring multi-region clients?

File: src/Signature/SignatureProvider.php [hunk 11]
```diff
@@ -109,6 +109,7 @@ public static function version()
     {
         return function ($version, $service, $region) {
             switch ($version) {
+                case 's3v4':
```
Comment (mtdowling) [inline src/Signature/SignatureProvider.php]: It might be a good idea to add a test to ensure s3v4 can be resolved. While it's dead simple right now, it would help in case of a future refactor.

File: tests/Endpoint/PartitionTest.php [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+            // partition endpoint setting
+            [$partition, 'us-east-1', 'baz', 's3v4'],
+            // service setting
+            [$partition, 'buzz', 'fizz', 'anonymous'],
+            // service endpoint setting
+            [$partition, 'pop', 'fizz', 's3v4'],
+            // no overrides
+            [$partition, 'puff', 'quux', 'v4'],
+            // unknown service
+            [$partition, 'us-east-1', 'iot', 'v4'],
```
Comment (mtdowling) [inline tests/Endpoint/PartitionTest.php]: How is this one resolving? "us-east-1" doesn't match the region explicitly nor do I see a region regex that it would match? Is this a test for an unknown service and and unknown region (e.g., fallback to assuming the aws partition as a last resort)? If so, maybe you should add a test that is just for unknown service, and one that is just for unknown region.

**Comment by mtdowling** [comment]
Looking good! Have you tested out the new endpoint resolution against the old pattern based system to make sure you're resolving them in the same way? I did that in Botocore by finding every permutation endpoints from the new system and compared it against what the old system would have resolved.

File: src/CommandPool.php (hunk 2 above)
Comment (mtdowling) [inline src/CommandPool.php]: Ah ok

**Comment by jeskew** [comment]
@mtdowling I didn't add a test for each service in each region, but the trickier cases were covered by `Aws\Test\PatternEndpointProviderTest::testResolvesEndpoints`. I moved those cases to the PartitionEnpointProviderTest.

**Comment by jeskew** [comment]
This PR has been updated with the latest version of endpoints.json. In addition, I've removed the `mapRegions` method to keep this change focussed.

/cc @mtdowling @xibz @cjyclaire

File: src/Endpoint/Partition.php [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        }
+
+        return false;
+    }
+
+    public function getRegionsForService($service)
+    {
+        if (isset($this->definition['services'][$service]['isRegionalized'])
+            && false === $this->definition['services'][$service]['isRegionalized']
+        ) {
```
Comment (cjyclaire) [inline src/Endpoint/Partition.php]: This first if statement seems to be the same with function resolveRegion($service, $region) despite that array is returned. While I can see that these 2 functions have different purpose, can you resolve the duplicates? Since $region parameter is only returned when the if check fails?

File: src/Endpoint/Partition.php [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        ];
+
+        $possibilities = array_intersect(
+            $supportedBySdk,
+            isset($data['signatureVersions'])
+                ? $data['signatureVersions']
+                : ['v4']
+        );
+
+        return array_pop($possibilities);
```
Comment (cjyclaire) [inline src/Endpoint/Partition.php]: While the logic look right, I'm little confused about the array operations going on there, is the $possibilities just has single element? If not, why just pop the last element of the result? If so, is there any privilege for using array_intersect for checking existence of an element in an array?

File: src/Endpoint/Partition.php [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+        $this->definition = $definition;
+    }
+
+    public function toArray()
+    {
+        return $this->definition;
+    }
+
+    public function offsetGet($offset)
+    {
```
Comment (cjyclaire) [inline src/Endpoint/Partition.php]: It is just a personal thing that I feel it is more nature and common to name the getter and setter like 'getOffset', 'setOffset', 'unsetOffset'.

File: tests/Endpoint/PartitionEndpointProviderTest.php [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+     *
+     * @param string $region
+     * @param string $partition
+     */
+    public function testResolvesPartitionsByRegion($region, $partition)
+    {
+        // Use the default endpoints file
+        $p = PartitionEndpointProvider::defaultProvider();
+        $this->assertSame($partition, $p->getPartitionFromRegion($region)['partition']);
+    }
```
Comment (cjyclaire) [inline tests/Endpoint/PartitionEndpointProviderTest.php]: In consideration for completeness and consistency, how about adding the test for unknown region and check whether default region is added as well?

File: src/Endpoint/Partition.php (hunk 13 above)
Comment (jeskew) [inline src/Endpoint/Partition.php]: That's a good point. I'll factor that out into a separate method.

File: src/Endpoint/Partition.php (hunk 14 above)
Comment (jeskew) [inline src/Endpoint/Partition.php]: I'm using `array_intersect` here to make sure that only signature versions that are supported by the SDK are pulled from the endpoints file. For example, Import/Export's signature versions are: `['v2', 'v4']`; the intersection of that against the SDK-supported signature versions will return an array with just one element: `v4`
```php
$supported = ['s3v4', 'v4', 'anonymous'];
$valid = array_intersect($supported, ['v2', 'v4']);
// $valid is [1 => 'v4']; 
```

If the service and SDK support more than one signature version, we should use them in the order of preference provided by the `$supportedBySdk` variable. I should be using `array_shift` instead of `array_pop` to enforce that.

File: src/Endpoint/Partition.php (hunk 15 above)
Comment (jeskew) [inline src/Endpoint/Partition.php]: These method signatures are defined by [`ArrayAccess`](http://php.net/manual/en/class.arrayaccess.php), so we're stuck with the names they have, unfortunately.

File: src/Endpoint/Partition.php (hunk 14 above)
Comment (cjyclaire) [inline src/Endpoint/Partition.php]: I see, make sense to me.

File: src/Endpoint/Partition.php (hunk 15 above)
Comment (cjyclaire) [inline src/Endpoint/Partition.php]: Okay, then it should be alright.

File: tests/Endpoint/PartitionEndpointProviderTest.php (hunk 16 above)
Comment (jeskew) [inline tests/Endpoint/PartitionEndpointProviderTest.php]: Will do!

**Comment by cjyclaire** [comment]
Looks good to me! :shipit:

**Comment by mtdowling** [comment]
There's one last test case to write to validate the endpoints against a list of known endpoints. Then 🚢

**Comment by jeskew** [comment]
@mtdowling I see your 🚢 and raise you a pile of documentation and tests.
